File: test/fixtures/swarm-tool-heartbeat-probe.py
# Copyright (C) 2026 Huawei Technologies Co., Ltd
# SPDX-License-Identifier: Apache-2.0
"""Real Swarm Model parser/watchdog probe; launched by swarm-tool-heartbeat-probe.mjs."""
import asyncio,json,sys
from openjiuwen.core.foundation.llm import Model,ModelClientConfig,ModelRequestConfig

async def main():
    config=json.loads(sys.stdin.read())
    model=Model(ModelClientConfig(client_provider='OpenAI',api_base=config['url'],api_key=config['token'],stream_idle_timeout=.5,stream_first_chunk_timeout=10,max_retries=0),ModelRequestConfig(model='test'))
    chunks=[]
    try:
        async for chunk in model.stream(messages=[{'role':'user','content':'test'}]): chunks.append(chunk)
    except Exception as exc:
        if config['case']=='stall':
            assert 'stream frame timeout' in str(exc),str(exc)
            print('RESULT '+json.dumps({'case':'stall','status':'correctly timed out'}));return
        raise
    assert config['case']!='stall','Real upstream silence must still time out'
    combined=chunks[0]
    for c in chunks[1:]:combined=combined+c
    calls=combined.tool_calls or []
    if config['case']=='truncated':
        assert not calls,calls
        assert 'output_limit:tool_calls_withheld' in combined.content
    else:
        assert len(calls)==1,calls
        assert json.loads(calls[0].arguments)=={'command':'private-command'}
        assert combined.content=='',combined.content
    print('RESULT '+json.dumps({'case':config['case'],'chunks':len(chunks),'tool_calls':len(calls),'status':'passed'}))
asyncio.run(main())

File: test/fixtures/swarm-tool-heartbeat-probe.mjs
// Copyright (C) 2026 Huawei Technologies Co., Ltd
// SPDX-License-Identifier: Apache-2.0
// Optional integration probe using the installed Swarm Python environment:
// SWARM_PYTHON=/path/to/swarm/.venv/bin/python node --import tsx test/fixtures/swarm-tool-heartbeat-probe.mjs
// No paid model requests. Tests actual gateway -> OpenAI SDK -> Swarm Model.stream watchdog.
// A 0.5s idle budget and 1s argument stream exercise the same boundary as the production 60s budget.
import {spawn} from 'node:child_process';
import {fileURLToPath} from 'node:url';
import {startModelGateway} from '../../services/api/src/agent-run/jiuwenswarm-model-gateway.ts';
const controller=new AbortController();
for(const scenario of ['complete','truncated','stall']){
 const streamer=async(_e,_s,_h,_t,_p,signal,callbacks)=>{
  for(let i=0;i<20;i++){
   await new Promise(r=>setTimeout(r,50));
   if(signal.aborted)throw new Error('aborted');
   callbacks.onProgress();
   callbacks.onToolCallDelta({index:0,id:'tool-1',name:'run_shell',arguments:'private-fragment'});
   if(scenario==='stall'){await new Promise(r=>setTimeout(r,1000));signal.throwIfAborted();}
  }
  const tool={id:'tool-1',name:'run_shell',args:{command:'private-command'}};
  return {truncated:scenario==='truncated',assistantMessage:{role:'assistant',content:''},toolCalls:[tool]};
 };
 const g=await startModelGateway({baseUrl:'http://unused',model:'test',apiProtocol:'anthropic-messages'},{maxRetries:0,maxTokens:1000,requestTimeoutMs:10000},controller.signal,streamer);
 try{
  const child=spawn(process.env.SWARM_PYTHON ?? 'python3',[fileURLToPath(new URL('./swarm-tool-heartbeat-probe.py', import.meta.url))],{stdio:['pipe','inherit','inherit'], timeout:30_000});
  child.stdin.end(JSON.stringify({url:g.url,token:g.token,case:scenario}));
  const code=await new Promise((resolve,reject)=>{child.on('error',reject);child.on('exit',resolve);});if(code!==0)throw new Error(`case ${scenario} exit ${code}`);
 }finally{await g.close();}
}
